src/Microsoft.Azure.Functions.Extensions.Mcp/Http/StreamableHttpRequestHandler.cs

```diff
@@ -2,18 +2,214 @@
+    }
+
+    private async ValueTask<IMcpClientSession<StreamableHttpTransport>?> GetOrCreateSessionAsync(HttpContext context, McpOptions mcpOptionsValue, bool stateless = true)
+    {
+        if (!stateless)
+        {
+            throw new NotSupportedException("Stateful sessions are not supported.");
+        }
+
+        var sessionId = context.Request.Headers[McpSessionIdHeaderName].ToString();
+
+        if (string.IsNullOrEmpty(sessionId))
+        {
+            return await StartNewSessionAsync(context, stateless);
+        }
+
+        if (!ClientStateManager.TryParseUriState(sessionId, out string? clientId, out string? instanceId, mcpOptionsValue.EncryptClientState))
+        {
+            await WriteJsonRpcErrorAsync(context, "Invalid client state.");
+
+            return null;
+        }
+
+        return await RebuildStatelessSessionAsync(context);
+    }
+
+    private async ValueTask<IMcpClientSession<StreamableHttpTransport>> StartNewSessionAsync(HttpContext context, bool stateless)
+    {
+        if (!stateless)
+        {
+            throw new NotSupportedException("Stateful mode is not supported.");
+        }
+
+        string clientId = Utility.EmptyId;
+        StreamableHttpTransport transport = new()
+        {
+            IsStateless = true,
+            SessionContext = new(clientId, instanceIdProvider.InstanceId),
+            OnInitRequestReceived = (t, p) =>
+                {
+                    var sessionId = Utility.CreateId();
+                    var clientState = ClientStateManager.FormatUriState(sessionId, instanceIdProvider.InstanceId, mcpOptions.Value.EncryptClientState);
+
+                    // Persist the session ID in the response header after receiving the initialize request.
+                    // TODO: Persist client information here from `initRequestParams.ClientInfo`
+                    // Do we need any additional client information? With that, if we're limiting to client info, we won't be able to provide
+                    // capabilities and other client details.
+                    t.SessionId = sessionId;
+                    context.Response.Headers[McpSessionIdHeaderName] = clientState;
+
+                    return ValueTask.CompletedTask;
+                }
+        };
+
+        // Create a new session with the transport.
+        // this session is currently being persisted in the session manager.
+        return await CreateSessionAsync(context, transport, clientId);
+    }
+
+    private async ValueTask<IMcpClientSession<StreamableHttpTransport>> RebuildStatelessSessionAsync(HttpContext context)
+    {
+        var sessionId = context.Request.Headers[McpSessionIdHeaderName].ToString();
+
+        if (!ClientStateManager.TryParseUriState(sessionId, out var clientId, out var instanceId, mcpOptions.Value.EncryptClientState))
+        {
+            await WriteJsonRpcErrorAsync(context, "Invalid client state.", StatusCodes.Status400BadRequest);
+            return null!;
```

Comment:
Should we make the return type nullable? `ValueTask<IMcpClientSession<StreamableHttpTransport>?>` like we have for GetOrCreateSessionAsync?

`private async ValueTask<IMcpClientSession<StreamableHttpTransport>?> GetOrCreateSessionAsync` ?